cmd/migrate/main.go

```diff
@@ -0,0 +1,161 @@
+	"k8s.io/client-go/util/retry"
+)
+
+var (
+	kubeconfig string
+	masterURL  string
+)
+
+func main() {
+	flag.StringVar(&kubeconfig, "kubeconfig", "", "Path to a kubeconfig. Only required if out-of-cluster.")
+	flag.StringVar(&masterURL, "master", "", "The address of the Kubernetes API server. Overrides any value in kubeconfig. Only required if out-of-cluster.")
+	flag.Parse()
+
+	cfg, err := clientcmd.BuildConfigFromFlags(masterURL, kubeconfig)
+	if err != nil {
+		glog.Fatalf("Error building kubeconfig: %v", err)
+	}
+
+	apiExtClient := apiextclient.NewForConfigOrDie(cfg)
+	clusterv1alpha1Client := clusterv1alpha1clientset.NewForConfigOrDie(cfg)
+	kubeClient := kubernetes.NewForConfigOrDie(cfg)
+
+	if err = migrateIfNecesary(kubeClient, apiExtClient, clusterv1alpha1Client, cfg); err != nil {
+		glog.Fatalf("Failed to migrate: %v", err)
+	}
+}
+
+func migrateIfNecesary(kubeClient kubernetes.Interface,
+	apiextClient apiextclient.Interface,
+	clusterv1alpha1Client clusterv1alpha1clientset.Interface,
+	config *restclient.Config) error {
+
+	_, err := apiextClient.ApiextensionsV1beta1().CustomResourceDefinitions().Get(downstreammachines.CRDName, metav1.GetOptions{})
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("failed to get crds: %v", err)
+	}
+
+	downstreamClient, err := downstreammachineclientset.NewForConfig(config)
+	if err != nil {
+		return fmt.Errorf("failed to create downstream machine client: %v", err)
+	}
+
+	if err = migrateMachines(kubeClient, downstreamClient, clusterv1alpha1Client); err != nil {
+		return fmt.Errorf("failed to migrate machines: %v", err)
+	}
+	if err = apiextClient.ApiextensionsV1beta1().CustomResourceDefinitions().Delete(downstreammachines.CRDName, nil); err != nil {
+		return fmt.Errorf("failed to delete downstream crd: %v", err)
+	}
+
+	return nil
+}
+
+func migrateMachines(kubeClient kubernetes.Interface,
+	downstreamClient downstreammachineclientset.Interface,
+	clusterv1alpha1Client clusterv1alpha1clientset.Interface) error {
+
+	// Get downstreamMachines
+	downstreamMachines, err := downstreamClient.MachineV1alpha1().Machines().List(metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to list downstream machines: %v", err)
+	}
+
+	// Convert the machine, create the new machine, delete the old one, wait for it to be absent
+	// We do this in one loop to avoid ending up having all machines in  both the new and the old format if deletion
+	// failes for whatever reason
+	for _, downstreamMachine := range downstreamMachines.Items {
+		convertedClusterv1alpha1Machine, err := conversions.ConvertV1alpha1DownStreamMachineToV1alpha1ClusterMachine(downstreamMachine)
+		if err != nil {
+			return fmt.Errorf("failed to convert machine %s: %v", downstreamMachine.Name, err)
+		}
+
+		createdClusterV1alpha1Machine, err := clusterv1alpha1Client.ClusterV1alpha1().Machines(convertedClusterv1alpha1Machine.Namespace).Create(convertedClusterv1alpha1Machine)
```

Comment:
> Then it is supposed to fail and try again. If it can not migrate the machine-controller will not come up as this runs in an initContainer. If that case ever occurs an operator has to check what the issue is.

since we can try and fail, we should do `Get` before, Am I right ?